File: xconf-angular-admin/src/main/webapp/app/xconf/settings/settingrule/settingrule.service.js
/* 
 * If not stated otherwise in this file or this component's Licenses.txt file the 
 * following copyright and licenses apply:
 *
 * Copyright 2018 RDK Management
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Author: Igor Kostrov
 * Created: 3/16/2016
 */
(function() {
    'use strict';

    angular
        .module('app.settingrule')
        .factory('settingRuleService', service);

    service.$inject=['utilsService', '$http', 'syncHttpService'];

    function service(utilsService, $http, syncHttpService) {
        var API_URL = 'setting/rule/';

        return {
            getAll: getAll,
            getRules: getRules,
            getRule: getRule,
            createRule: createRule,
            updateRule: updateRule,
            deleteRule: deleteRule,
            updateSyncEntities: updateSyncEntities,
            createSyncEntities: createSyncEntities
        };

        function getAll() {
            return $http.get(API_URL);
        }

        function getRules(pageNumber, pageSize, searchContext) {
            var url = API_URL + 'filtered?pageNumber=' + pageNumber + '&pageSize=' + pageSize;
            return $http.post(url, searchContext);
        }

        function getRule(id) {
            return $http.get(API_URL + id);
        }

        function createRule(firmwareRule) {
            return $http.post(API_URL, firmwareRule);
        }

        function updateRule(firmwareRule) {
            return $http.put(API_URL, firmwareRule);
        }

        function deleteRule(id) {
            return $http.delete(API_URL + id);
        }

        function updateSyncEntities(firmwareRules) {
            var requests = utilsService.generateRequestList(firmwareRules, {url: API_URL + 'entities', method: 'PUT'});
           return requests && requests.length ? syncHttpService.http(requests) : null;
        }

        function createSyncEntities(firmwareRules) {
            var requests = utilsService.generateRequestList(firmwareRules, {url: API_URL + 'entities', method: 'POST'});
            return requests && requests.length ? syncHttpService.http(requests) : null;
        }
    }
})();

File: xconf-angular-admin/src/main/java/com/comcast/xconf/admin/service/setting/SettingRuleService.java
/* 
 * If not stated otherwise in this file or this component's Licenses.txt file the 
 * following copyright and licenses apply:
 *
 * Copyright 2018 RDK Management
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Author: Igor Kostrov
 * Created: 3/21/2016
*/
package com.comcast.xconf.admin.service.setting;

import com.comcast.hesperius.dataaccess.core.dao.ISimpleCachedDAO;
import com.comcast.xconf.admin.validator.setting.SettingRuleValidator;
import com.comcast.xconf.logupload.settings.SettingRule;
import com.comcast.xconf.permissions.DcmPermissionService;
import com.comcast.xconf.permissions.PermissionService;
import com.comcast.xconf.search.PredicateManager;
import com.comcast.xconf.search.SearchFields;
import com.comcast.xconf.shared.service.AbstractApplicationTypeAwareService;
import com.comcast.xconf.util.RuleUtil;
import com.comcast.xconf.validators.IValidator;
import com.google.common.base.Predicate;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

@Component
public class SettingRuleService extends AbstractApplicationTypeAwareService<SettingRule> {

    @Autowired
    private ISimpleCachedDAO<String, SettingRule> settingRuleDAO;

    @Autowired
    private SettingRuleValidator validator;

    @Autowired
    private PredicateManager predicateManager;

    @Autowired
    private DcmPermissionService permissionService;

    @Override
    public PermissionService getPermissionService() {
        return permissionService;
    }

    @Override
    public ISimpleCachedDAO<String, SettingRule> getEntityDAO() {
        return settingRuleDAO;
    }

    @Override
    public IValidator<SettingRule> getValidator() {
        return validator;
    }

    @Override
    protected void normalizeOnSave(SettingRule entity) {
        RuleUtil.normalizeConditions(entity.getRule());
    }

    @Override
    protected List<Predicate<SettingRule>> getPredicatesByContext(Map<String, String> context) {
        List<Predicate<SettingRule>> predicates = new ArrayList<>();
        if (context == null || context.isEmpty()) {
            return predicates;
        }

        Predicate<SettingRule> xRulePredicate = predicateManager.getXRulePredicate(context);
        if (xRulePredicate != null) {
            predicates.add(xRulePredicate);
        }
        if (context.containsKey(SearchFields.NAME)) {
            predicates.add(predicateManager.new XRuleNamePredicate<SettingRule>(context.get(SearchFields.NAME)));
        }
        return predicates;
    }
}
